rhostck.1
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.13)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "rhostck 1"
.TH rhostck 1 " " "" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
rhostck \- tell rblsmtpd to block mails from hosts with suspicious names
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBrhostck rblsmtpd\fR ...
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBrhostck\fR is supposed to run under \fB\f(BItcpserver\fB\|(1)\fR and to execute
\&\fB\f(BIrblsmtpd\fB\|(1)\fR.
.PP
\&\fBrhostck\fR checks the \f(CW$TCPREMOTEHOST\fR environment variable. Based on
the name of the remote host \fBrhostck\fR may decide to set the \f(CW$RBLSMTPD\fR
environment variable in order to tell \fBrblsmtpd\fR to reject the incoming
\&\s-1SMTP\s0 delivery request. By default \fBrhostck\fR instructs \fBrblsmtpd\fR to
reject remote hosts without a hostname.
.PP
In the environment variable \f(CW$RHOSTCK_DENYPARTS\fR a space separated list
of name parts of a hostname can be defined, that typically occur in
hostnames of dynamic lines. \fBrhostck\fR splits the remote hostname at the
dots and looks at each part in front of the penultimate dot. \f(CW$RBLSMTPD\fR
will be set with an error message, if at least one such part exists and
starts with one of the parts given in \f(CW$RHOSTCK_DENYPARTS\fR.
.PP
If the environment variable \f(CW$RHOSTCK_IPHOSTNAME\fR is filled with a valid
value, the part of remote hostname in front of the penultimate dot is
checked to contain the remote host's ip address (or part of it). Valid
values are \f(CW2\fR, \f(CW3\fR and \f(CW4\fR, which defines the number of bytes from the
binary \s-1IP\s0 address from \s-1LSB\s0 to \s-1MSB\s0 (or numbers of the dot-decimal notation
from right to left) to be used for comparing. The (partial) \s-1IP\s0 address is
searched in the hostname in decimal and hexdecimal form; the numbers may
be separated by dots or hyphens, or without separators for decimal numbers
each number is expected to be three bytes long and left-zero-padded,
whereas the hexadecimal form uses the usual two nibbles for a byte.
All comparisons are also done in reverse order. If the (partial) remote \s-1IP\s0
address is found in the remote hostname in any form described above,
\&\f(CW$RBLSMTPD\fR will be set with an error message.
.PP
The error message for \fBrblsmtpd\fR is taken from the content of
\&\f(CW$RHOSTCK_DENYMSG\fR. If this is not set or emtpy the default text is
\&\f(CW"\-Suspicious hostname \- I don\*(Aqt accept mails from hosts on dynamic IP addresses."\fR;
it starts with a hyphen to instruct \fBrblsmtpd\fR to terminate the
connection with a 553 permanent error.
Eventually \fBrhostck\fR executes the command given as its first argument,
which is \fBrblsmtpd\fR normally.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fItcpserver\fR\|(1), \fIrblsmtpd\fR\|(1)
.PP
http://cr.yp.to/ucspi\-tcp.html <http://cr.yp.to/ucspi-tcp.html>


CHANGES
0.5	2010-04-25
- Check if remote IP address is part of the remote hostname.

0.4	2010-04-20
- Look at each part in front of the second last dot.

0.3	2010-04-18
- Making rblsmtpd message configurable.
- Default rblsmtpd message now starts with a hyphen for a 553 error.
- Bug fix: Installation was broken.
- Documentation enhancements.
- Adding an RPM SPEC file.
- Adding a manpage.

0.2	2010-04-17
- Switching to DJB style.
- Set of hostname parts is configurable.
- Documentation.

0.1	2008-03-06
- Initial version, blocking from a fix set of hostname parts.


rhostck.c
#include <sys/types.h>
#include <sys/stat.h>
#include "buffer.h"
#include "case.h"
#include "env.h"
#include "fmt.h"
#include "hasinline.h"
#include "ip4.h"
#include "pathexec.h"
#include "str.h"
#include "stralloc.h"
#include "strerr.h"
#include "version.h"

#define FATAL "rhostck: fatal: "
#define DENYMSG "-Suspicious hostname - I don't accept mails from hosts on dynamic IP addresses."

const char version[] = "rhostck " VERSION;

char *tcpremoteip;
char *tcpremotehost;
char ip[4] = { 0, 0, 0, 0 };
char strnum[FMT_ULONG];
static stralloc relevant_hostname_part;
char ip_permuted[IP4_FMT];

void nomem(void)
{
    strerr_die2x(111, FATAL, "out of memory");
}

static inline char tohex(char c) {
    return c >= 10 ? c-10 + 'a' : c + '0';
}

/*
 * Check the given character.
 * Return 0 if character is between '0' and '9', else return 1.
 */
int not_numeric(unsigned char c)
{
    register unsigned char x = c - '0';

    return (x > '9' - '0');
}

/*
 * Check the given character. Return 0 if character is between 'A' and 'Z'
 * or 'a' and 'z' or '0' and '9'. Return 1 otherwise.
 */
int not_alphanumeric(unsigned char c)
{
    register unsigned char x;

    x = c - 'A';
    if (x <= 'Z' - 'A') return 0;
    x = c - 'a';
    if (x <= 'z' - 'a') return 0;

    return not_numeric(c);
}

/*
 * Check the given character. Return 0 if character is between 'A' and 'F'
 * or 'a' and 'z' or '0' and '9'. Return 1 otherwise.
 */
int not_hexanumeric(unsigned char c)
{
    register unsigned char x;

    x = c - 'A';
    if (x <= 'F' - 'A') return 0;
    x = c - 'a';
    if (x <= 'f' - 'a') return 0;

    return not_numeric(c);
}

/*
 * ip_start returns 1 if 2nd argument is a prefix of the first, 0 otherwise.
 * Characters '.' and '-' are treated as equal!
 */
int ip_start(register const char *s,register const char *t)
{
  register char x;
  register char y;

  for (;;) {
    x = *t++; if (!x) return 1; y = *s++; if (y == '-') y = '.'; if (x != y) return 0;
    x = *t++; if (!x) return 1; y = *s++; if (y == '-') y = '.'; if (x != y) return 0;
  }
}
/*
 */
int find_ip(const char *name, const char *ip, int (*not_chartype)(unsigned char))
{
    char c = *ip;
    size_t p, pp;
    size_t iplen = str_len(ip);

#if 0
buffer_puts(buffer_1,"hostname part: ");
buffer_puts(buffer_1,name);
buffer_puts(buffer_1,"\n");
buffer_puts(buffer_1,"ip part: ");
buffer_puts(buffer_1,ip);
buffer_putsflush(buffer_1,"\n");
#endif

    for (p = 0; name[p]; ++p) {
	pp = str_chr(name + p, c);
	p += pp;
	if (!name[p]) break;
	if (ip_start(name + p, ip)) {
	    if ((!p || (*not_chartype)(name[p-1])) && (*not_chartype)(name[p + iplen]))
		return 1;
	}
    }
    return 0;
}

/*
 * Compare all tokens of denyparts with the begining of start.
 * Return 1 if any of the tokens equals start and the next character
 * of start is not alphanumeric.
 */
unsigned int find_deny_token(const char *denyparts, const char *start)
{
    size_t i = 0, len = 0;
    const char *token = denyparts;

    for (;;) {
	if (!token[i])
	    len = i;
	else if (token[i] == ' ') {
	    len = i;
	    do { ++i; } while (token[i] == ' ');
	}
	else {
	    ++i;
	}
	if (len) {
	    if (case_diffb(start, len, token) == 0 && not_alphanumeric(start[len]))
		return 1;
	    token += i;
	    i = len = 0;
	}
	if (!*token) break;
    }

    return 0;
}

/*
 * Check TCPREMOTEHOST environment variable.
 * Return 1 if we consider this host's name an untrustfull source.
 */
int fishy_remotehost(void)
{
    const char *t, *denyparts;
    int start;

    denyparts = env_get("RHOSTCK_DENYPARTS");
    if (!denyparts) return 0;

    while (*denyparts == ' ') ++denyparts;	// skip leading spaces

    /*
     * Split relevant part of remote hostname at dots and compare
     * beginning of each part with all tokens in $RHOSTCK_DENYPARTS.
     * Return true if any part begins with any token, else false.
     */

    for (start = 1, t = relevant_hostname_part.s; *t; ++t)
	if (*t == '.')
	    start = 1;
	else
	    if (start) {
		if (find_deny_token(denyparts, t))
		    return 1;
		start = 0;
	    }

    return 0;	// one dot only or no suspicious hostname parts detected
}

/*
 */
int ip_in_hostname(void)
{
    const char *tuples;
    char *s;
    unsigned long n;	// 2, 3 or 4
    size_t pos;
    unsigned long numlen;
    unsigned int i;
    const char *name = relevant_hostname_part.s;
    unsigned char c;

    s = env_get("RHOSTCK_IPHOSTNAME");

    if (!s)
	return 0;

    if (!scan_ulong(s, &n) || n < 2 || n > 4)
	return 0;

    // try to find the verbatim IP address (part) in hostname
    s = tcpremoteip;
    for (i = 4; i > n; --i)
	s += str_chr(s, '.') + 1;

    if (find_ip(name, s, not_numeric)) return 1;

    tuples = ip + (4 - n);

    // reverse ip
    s = ip_permuted;
    i = n - 1;
    for (;;) {
	s += fmt_ulong(s, (unsigned long)(unsigned char)tuples[i]);
	if (!i--) break;
	*s++ = '.';
    }
    *s++ = 0;

    if (find_ip(name, ip_permuted, not_numeric)) return 1;

    // ip with each number pre-padded with 0s to 3 bytes w/o stop bytes
    s = ip_permuted;
    for (i = 0; i < n; ++i) {
	c = tuples[i];
	pos = 3 - fmt_ulong(0, (unsigned long)c);
	while (pos--) *s++ = '0';
	s += fmt_ulong(s, (unsigned long)c);
    }
    *s = 0;

    if (find_ip(name, ip_permuted, not_numeric)) return 1;

    // ip with each number pre-padded with 0s to 3 bytes w/o stop bytes reversed
    s = ip_permuted;
    i = n - 1;
    do {
	c = tuples[i];
	pos = 3 - fmt_ulong(0, (unsigned long)c);
	while (pos--) *s++ = '0';
	s += fmt_ulong(s, (unsigned long)c);
    } while (i--);
    *s = 0;

    if (find_ip(name, ip_permuted, not_numeric)) return 1;

    // hex ip
    s = ip_permuted;
    for (i = 0; i < n; ++i) {
	c = tuples[i];
	*s++ = tohex(c >> 4);
	*s++ = tohex(c & 15);
    }
    *s = 0;

    if (find_ip(name, ip_permuted, not_hexanumeric)) return 1;

    // hex ip reversed
    s = ip_permuted;
    i = n - 1;
    do {
	c = tuples[i];
	*s++ = tohex(c >> 4);
	*s++ = tohex(c & 15);
    } while (i--);
    *s = 0;

    if (find_ip(name, ip_permuted, not_hexanumeric)) return 1;

    // hex ip with dots/dashes
    s = ip_permuted;
    i = 0;
    for (;;) {
	c = tuples[i];
	*s++ = tohex(c >> 4);
	*s++ = tohex(c & 15);
	if (++i >= n) break;
	*s++ = '.';
    }
    *s = 0;

    if (find_ip(name, ip_permuted, not_hexanumeric)) return 1;

    // hex ip reversed with dots/dashes
    s = ip_permuted;
    i = n - 1;
    for (;;) {
	c = tuples[i];
	*s++ = tohex(c >> 4);
	*s++ = tohex(c & 15);
	if (!i--) break;
	*s++ = '.';
    }
    *s = 0;

    if (find_ip(name, ip_permuted, not_hexanumeric)) return 1;

    return 0;
}

int main(int argc, char** argv)
{
    const char *t, *dot1, *dot2;

    if (argc < 2)
	strerr_die1x(100, "rhostck: usage: rhostck program");

    tcpremoteip = env_get("TCPREMOTEIP");
    if (!tcpremoteip) tcpremoteip = "";

    if (!ip4_scan(tcpremoteip, ip))
	strerr_die2x(111, FATAL, "TCPREMOTEIP not set or not an IPv4 address");

    tcpremotehost = env_get("TCPREMOTEHOST");
    if (!tcpremotehost) tcpremotehost = "";

    // find dots in TCPREMOTEHOST
    // no dots: "localhost" -> block
    // one dot only: "name.tld" -> pass
    // two or more dots: "host.name.tld"
     
    t = tcpremotehost;
    dot1 = dot2 = 0;
    for (;;) {
	if (!*t) break;
	if (*t == '.') {
	    dot1 = dot2;
	    dot2 = t;
	}
	++t;
    }

    if (dot1) {
	// save string in front of penultimate dot.
	if (!stralloc_copyb(&relevant_hostname_part, tcpremotehost, (dot1 - tcpremotehost)))
	    nomem();
	if (!stralloc_0(&relevant_hostname_part)) nomem();
    }


    // !dot2 means: no dot exists in remote hostname -> bad hostname
    // dot1 means: at least two dots in gostname
    if (!dot2 || (dot1 && (fishy_remotehost() || ip_in_hostname()))) {
	char *rblmsg = env_get("RHOSTCK_DENYMSG");

	if (!rblmsg || !*rblmsg)
	    rblmsg = DENYMSG;

        pathexec_env("RBLSMTPD", rblmsg);
    }

    pathexec(argv + 1);

    strerr_die4sys(111, FATAL, "unable to start ", argv[1], ": ");
}
